src/piadm.sh

```diff
@@ -195,6 +196,104 @@ DEFAULT_URL_PREFIX=https://us-central.manta.mnx.io/Joyent_Dev/public/SmartOS/
 # Default path for piadm's configuration
 PIADM_CONF=/var/piadm/piadm.conf
 
+# fetch_checksum
+#
+# Fetches a MD5 hash using a platform file as the key.
+#
+# Arguments:
+#   $1 - URL from where to fetch the PI from. The name of the PI
+#   	 is expected to be the last path component of the URL,
+#   	 or the literal string "latest".
+#
+# Environment Variables:
+#   PIADM_NO_CHECKSUM     - If set to 1, skips checksum validation.
+#   PIADM_CHECKSUM_URL    - If set, overrides the default md5sums.txt URL
+#   					  and uses the value provided.
+#
+# Notes:
+#   - If "latest" is used, up to 1024 PIs will be listed from ${URL_PREFIX}.
+#     The last one is assumed to be the latest.
+#   - On error, this function exits with code 1, following PIADM(8)
+#     convention: the error indicates a failure, but no changes were made
+#     to the system.
+checksum_platform=""
+fetch_checksum() {
+	if [[ -z "$checksum_platform" ]]; then
+		local platform_file
+		local stamp
+		if [ "$1" != "latest" ]; then
+			IFS=' ' read -r stamp platform_file <<<\
+			$(echo "$1" | rev | cut -d'/' -f1,2 | rev | tr -s "/" " ")
+		else
+			#latest is a special case, we need find out stamp
+			stamps_url="${URL_PREFIX%/}?limit=1024"
+			stamp=$("${CURL[@]}" "${stamps_url}" |\
+			json -ga -c "this.name.match(/Z$/)" | json -ag name |\
+				sort  | tail -1; exit ${PIPESTATUS[0]})
+			code=$?
+			if [[ $code -ne 0 ]]; then
+				eecho "Curl failed fetching PI data from ${stamps_url}"
+				return 1
+			fi
+			platform_file="smartos-${stamp}.iso"
+		fi
+		if [[ -z ${PIADM_CHECKSUM_URL} ]];then
+			local checksum_url="${URL_PREFIX}${stamp}/md5sums.txt"
```

Comment:
Yep, THIS is where we keep things hardcoded for now.  Good job! (Can resolve this upon reading.)